rpc/handlers.go

```diff
@@ -1242,6 +1242,64 @@ func (h *Handler) TraceTransaction(hash felt.Felt) (json.RawMessage, *jsonrpc.Er
 	return trace, nil
 }
 
+func (h *Handler) SimulateTransactions(id BlockID, transactions []BroadcastedTransaction,
+	simulationFlags []SimulationFlag,
+) ([]SimulatedTransaction, *jsonrpc.Error) {
+	if len(simulationFlags) > 0 {
+		return nil, jsonrpc.Err(jsonrpc.InvalidParams, "Simulation flags are not supported")
+	}
+
+	state, closer, err := h.stateByBlockID(&id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	defer h.callAndLogErr(closer, "Failed to close state in starknet_simulateTransactions")
+
+	header, err := h.blockHeaderByID(&id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	var (
+		txns    []core.Transaction
+		classes []core.Class
+
+		result []SimulatedTransaction
+	)
+	for i := range transactions {
+		transaction := transactions[i]
+		tx, class, paidFeeOnL1, err := adaptBroadcastedTransaction(&transaction, h.network)
+		if err != nil {
+			// todo or InvalidParams ?
+			return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+		}
+
+		txns = append(txns, tx)
+		classes = append(classes, class)
+
+		paidFeesOnL1 := []*felt.Felt{paidFeeOnL1}
+		gasConsumed, traces, err := h.vm.Execute(txns, classes, header.Number, header.Timestamp,
```

Comment:
Why are calling `vm.Execute()` for each txn?
Shouldnt we just call it once for the entire txn set?

Reply:
My bad, fixed